### PR TITLE
Implement CanvasRenderingContext2d.fillText's "unimplemented" message

### LINKED ISSUE #11681: Implement CanvasRenderingContext2d.fillText
Spec: https://html.spec.whatwg.org/multipage/scripting.html#dom-context-2d-filltext
Code: `components/script/dom/canvasrenderingcontext2d.rs`, `components/script/dom/weibidls/CanvasRenderingContext2D.webidl`, `components/canvas/canvas_paint_thread.rs`

(I hope I identified correctly the code files - followed the style in #5411)

### PR DESCRIPTION
**Pull request by BrunoBernardino**
Basic skeleton for implementing CanvasRenderingContext2d.fillText,  only adding methods and parameters in the right place, and a basic test, with some `println!()`.

This is only the beginning. It were my first couple of hours looking at Rust and Servo.

However, I have _no clue_ how to get the text to render now (basically go from the `println!()` to something else).

It's also possible I messed something up with the `DOMString.parse()` but not entirely sure.

I'm doing this PR as a starting point to get help and learn more about this, _or_ at least maybe save someone some time while implementing this, if no one's able to take the time and show me where/how.

Because it's still a work-in-progress, I'm leaving the boxes below unchecked (even though there are no errors).

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #11681 and #17963

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18020)

### CODE REVIEW

**Comment by emilio** [comment]
So first of all, this is awesome, thank you! :)

It's a bit unclear to me how easy it is to fill the missing pieces here. For the current 2d canvas APIs for now we use [bindings to azure](https://github.com/servo/rust-azure).

So, in theory, the only thing that we have to do is to expose the text drawing APIs azure has, and use them from here, with the current parameters (that is, font-size, etc).

That being said, I suspect this is not an easy task though. As far as I understand, all the text caches in azure aren't thread-safe, so it'd need a fair amount of work.

It'd be nice to try to use another back-end, but that's also a huge amount of work. If we want to implement this properly, I'd rather discuss with people what's the best long-term fix for this. I know @asajeffrey has been fiddling with Canvas2D for a while now, so maybe he has ideas. @glennw may also have an opinion on what's the best way to support text stuff in Canvas2D.

**Comment by asajeffrey** [comment]
I've talked to @pcwalton about using a back-end based on webrender + pathfinder, so I'm not sure how much effort we should put in to updating the current azure back end.

**Comment by MortimerGoro** [comment]
It may be interesting to provide a mock implementation for fillText or other missing Canvas2D methods before the long term webrender/pathfinder solution lands (any ETA on this? I'm assuming that it can take some time). Some WebGL demos that I try crash due to those missing methods and usually is to render a sprite not related to the game itself (e.g. the fps stats).

I'd prefer some unimplemented warnings in the console instead of a crash in those cases. Also a lot of WebGL tests are exiting early because of the missing fillText() function, it would be good to enable them.

What do you think?

**Comment by jdm** [comment]
@MortimerGoro I am confused how a missing JS method could cause a crash/panic in Servo. Do you mean that the tests report a JS exception?

**Comment by MortimerGoro** [comment]
@jdm Yes, I meant that JavaScript/game execution "crashes" with a uncaught JS exception, not the servo process.

**Comment by jdm** [comment]
I would be open to adding stubs for missing canvas APIs that are hidden behind a preference.

**Comment by BrunoBernardino** [comment]
Can you elaborate on that @jdm ? Basically change my `console.log`s above to better `console.warn` and where would the preference be? (and how could I add it)?

**Comment by jdm** [comment]
Replace them with `error!(...)` instead; add a `[Pref="dom.canvas-text.enabled"]` to the CanvasRenderingContext2D.webidl file; add the new preference to `resources/prefs.json`.

**Comment by BrunoBernardino** [comment]
Not sure if this is what you were thinking about, and it doesn't really fix any of the mentioned issues, though, because they would require an actual implementation.

I do understand, as per @MortimerGoro's comment that it's fine and this will somewhat improve things.

screen shot 2017-08-19 at 18 15 51

screen shot 2017-08-19 at 18 21 01

Please let me know if I'm missing any other sort of unit tests or something.

File: components/script/dom/canvasrenderingcontext2d.rs [hunk 1]
```diff
@@ -800,6 +800,13 @@ impl CanvasRenderingContext2DMethods for CanvasRenderingContext2D {
         receiver.recv().unwrap()
     }
 
+    // https://html.spec.whatwg.org/multipage/#dom-context-2d-filltext
+    fn FillText(&self, text: DOMString, x: f64, y: f64, max_width: Option<f64>) {
+        let parsed_text: String = text.parse().unwrap();
```
Comment (emilio) [inline components/script/dom/canvasrenderingcontext2d.rs]: This doesn't need `parse()`, you should be able to do `into()`, or if it doesn't exist, you may as well add it.

Comment (BrunoBernardino) [inline components/script/dom/canvasrenderingcontext2d.rs]: Thank you @emilio! I'll do that!

Comment (BrunoBernardino) [inline components/script/dom/canvasrenderingcontext2d.rs]: Oh. I'm afraid `DOMString` doesn't have `into()` and I'm not sure how to add it to `components/script/dom/bindings/str.rs`, can you help?

Comment (BrunoBernardino) [inline components/script/dom/canvasrenderingcontext2d.rs]: Apparently just `.into()` works. I was trying `.into().unwrap()`. Thanks!

File: resources/prefs.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "dom.bluetooth.enabled": false,
   "dom.bluetooth.testing.enabled": false,
-  "dom.canvas-text.enabled": false,
+  "dom.canvas-text.enabled": true,
```
Comment (jdm) [inline resources/prefs.json]: Please revert this change. We do not want this to be exposed by default, since it is incomplete. Anybody that needs to run tests with this API present can enable it explicitly.

Comment (BrunoBernardino) [inline resources/prefs.json]: Sounds good! I thought the goal here was to prevent the JS Exception by default.

**Comment by BrunoBernardino** [comment]
Alright, I think that's it!

**Comment by jdm** [comment]
```
./resources/prefs.json:None: Unordered key (found dom.compositionevent.enabled before dom.canvas-text.enabled)
```
Otherwise, these changes look good. Could you [squash](https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/) all of the commits into one?

**Comment by jdm** [comment]
@bors-servo: r+
Thanks!

**Comment by jdm** [comment]
@bors-servo: retry
* #13342
